pkg/controller/user-cluster-controller-manager/application-installation-controller/controller.go

```diff
@@ -112,18 +128,116 @@ func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
 	}
 
 	log.Debug("Processed")
 	return reconcile.Result{}, err
 }
 
 func (r *reconciler) reconcile(ctx context.Context, log *zap.SugaredLogger, appInstallation *appkubermaticv1.ApplicationInstallation) error {
-	// to implements logic in next PR
+	// handling deletion
+	if !appInstallation.DeletionTimestamp.IsZero() {
+		if err := r.handleDeletion(ctx, log, appInstallation); err != nil {
+			return fmt.Errorf("handling deletion of application installation: %w", err)
+		}
+		return nil
+	}
+
+	if err := kuberneteshelper.TryAddFinalizer(ctx, r.userClient, appInstallation, appkubermaticv1.ApplicationInstallationCleanupFinalizer); err != nil {
+		return fmt.Errorf("failed to add finalizer: %w", err)
+	}
+
+	appHasBeenInstalled := appInstallation.Status.ApplicationVersion != nil
+
+	// get applicationDefition. if it can not be found there is 2 cases:
+	//   1) KKP admin has remove the applicationDefition and we have to remove the corresponding ApplicationInstallation(s)
+	//   2) User made a mistake or applicationDefinition has not been sync yet on this seed. so  we just notify the user.
+	applicationDef := &appkubermaticv1.ApplicationDefinition{}
+	if err := r.seedClient.Get(ctx, types.NamespacedName{Name: appInstallation.Spec.ApplicationRef.Name}, applicationDef); err != nil {
+		if apierrors.IsNotFound(err) {
+			if appHasBeenInstalled {
+				r.traceWarning(appInstallation, log, applicationDefinitionRemovedEvent, fmt.Sprintf("ApplicationDefinition '%s' has been deleted, removing applictionInstallation", applicationDef.Name))
+				return r.userClient.Delete(ctx, appInstallation)
+			} else {
+				return fmt.Errorf("ApplicationDefinition '%s' does not exist. can not install application", applicationDef.Name)
+			}
+		}
+		return err
+	}
+
+	if !applicationDef.DeletionTimestamp.IsZero() {
+		r.traceWarning(appInstallation, log, applicationDefinitionDeletingEvent, fmt.Sprintf("ApplicationDefinition '%s' is being deleted,  removing applictionInstallation", applicationDef.Name))
+		return r.userClient.Delete(ctx, appInstallation)
+	}
+
+	// get aplicationVersion. if it can not be found there is 2 cases:
+	//   1) KKP admin has remove the applicationVersion and we have to remove the corresponding ApplicationInstallation(s)
+	//   2) User made a mistake or applicationDefinition has not been sync yet on this seed. so  we just notify the user.
+	appVersion := &appkubermaticv1.ApplicationVersion{}
+	if err := r.getApplicationVersion(appInstallation, applicationDef, appVersion); err != nil {
+		if appHasBeenInstalled {
+			r.traceWarning(appInstallation, log, applicationVersionRemovedEvent, fmt.Sprintf("applicationVersion: '%s' has been deleted. removing Application", appInstallation.Spec.ApplicationRef.Version))
+			return r.userClient.Delete(ctx, appInstallation)
+		} else {
+			return fmt.Errorf("applicationVersion: '%s' does not exist. can not install application", appInstallation.Spec.ApplicationRef.Version)
```

Comment:
This `else` is for the `appHasBeenInstalled` rather than the version not existing right ?

Reply:
Let's say we have this ApplicationDefinition and ApplicationInstallation installed:
```
apiVersion: apps.kubermatic.k8c.io/v1
kind: ApplicationDefinition
metadata:
  name: app-def-1
spec:
  description: "appp 1 description"
  versions:
    - version: "1.0.0"
      constraints:
        k8sVersion: ">= 1.19.0"
        kkpVersion: ">= 2.0"
      template:
        method: helm
        source:
          helm:
            chartName: "aChartName"
            chartVersion: "aChartVersion"
            url: "https://localhost"
    - version: "2.0.0"
      constraints:
        k8sVersion: ">= 1.19.0"
        kkpVersion: ">= 2.0"
      template:
        method: helm
        source:
          helm:
            chartName: "aChartName"
            chartVersion: "aChartVersion"
            url: "https://localhost"
```

```
apiVersion: apps.kubermatic.k8c.io/v1
kind: ApplicationInstallation
metadata:
  name: example-v1
spec:
  targetNamespace: foo
  createNamespace: true
  applicationRef:
    name: app-def-1
    version: 1.0.0
```

now KKP admin remove version `1.0.0` from applicationDefinition
```
apiVersion: apps.kubermatic.k8c.io/v1
kind: ApplicationDefinition
metadata:
  name: app-def-1
spec:
  description: "appp 1 description"
  versions:
    - version: "2.0.0"
      constraints:
        k8sVersion: ">= 1.19.0"
        kkpVersion: ">= 2.0"
      template:
        method: helm
        source:
          helm:
            chartName: "aChartName"
            chartVersion: "aChartVersion"
            url: "https://localhost"
```
 Consequently, we have to remove all applications using this version. But how can we differentiate an applicationInstallation that references a version that never existed and a version that has been removed? 
To do that we check if the application has been installed at least once. if so that means the version has been removed.